spras/config/schema.py

```diff
@@ -0,0 +1,168 @@
+    euclidean = 'euclidean'
+    manhattan = 'manhattan'
+    cosine = 'cosine'
+
+class MlAnalysis(BaseModel):
+    include: bool
+    aggregate_per_algorithm: bool = False
+    components: int = 2
+    labels: bool = True
+    kde: bool = False
+    remove_empty_pathways: bool = False
+    linkage: MlLinkage = MlLinkage.ward
+    metric: MlMetric = MlMetric.euclidean
+
+    model_config = ConfigDict(extra='forbid')
+
+class EvaluationAnalysis(BaseModel):
+    include: bool
+    aggregate_per_algorithm: bool = False
+
+    model_config = ConfigDict(extra='forbid')
+
+class Analysis(BaseModel):
+    summary: SummaryAnalysis = SummaryAnalysis(include=False)
+    cytoscape: CytoscapeAnalysis = CytoscapeAnalysis(include=False)
+    ml: MlAnalysis = MlAnalysis(include=False)
+    evaluation: EvaluationAnalysis = EvaluationAnalysis(include=False)
+
+    model_config = ConfigDict(extra='forbid')
+
+
+# The default length of the truncated hash used to identify parameter combinations
+DEFAULT_HASH_LENGTH = 7
+
+def label_validator(name: str):
+    """
+    A validator takes in a label
+    and ensures that it's alphanumeric (with underscores).
+    """
+    label_pattern = r'^\w+$'
+    def validate(label: str):
+        if not bool(re.match(label_pattern, label)):
+            raise ValueError(f"{name} label '{label}' contains invalid values. {name} labels can only contain letters, numbers, or underscores.")
+        return label
+    return validate
+
+class ContainerFramework(CaseInsensitiveEnum):
+    docker = 'docker'
+    # TODO: add apptainer variant once #260 gets merged
+    singularity = 'singularity'
+    dsub = 'dsub'
+
+class ContainerRegistry(BaseModel):
+    base_url: str
+    owner: str = Field(description="The owner or project of the registry")
+
+    model_config = ConfigDict(extra='forbid')
+
+class AlgorithmParams(BaseModel):
+    include: bool
+    directed: Optional[bool] = None
+
+    # TODO: use array of runs instead. We currently rely on the
+    # extra parameters here to extract the algorithm parameter information,
+    # which is why this deviates from the usual ConfigDict(extra='forbid').
+    model_config = ConfigDict(extra='allow')
+
+class Algorithm(BaseModel):
+    name: str
+    params: AlgorithmParams
+
+    model_config = ConfigDict(extra='forbid')
+
+class Dataset(BaseModel):
+    # We prefer AfterValidator here to allow pydantic to run its own
+    # validation & coercion logic before we check it against our own
+    # requirements
+    label: Annotated[str, AfterValidator(label_validator("Dataset"))]
+    node_files: list[str]
+    edge_files: list[str]
+    other_files: list[str]
+    data_dir: str
+
+    model_config = ConfigDict(extra='forbid')
+
+class GoldStandard(BaseModel):
+    label: Annotated[str, AfterValidator(label_validator("Gold Standard"))]
+    node_files: list[str]
+    data_dir: str
+    dataset_labels: list[str]
+
+    model_config = ConfigDict(extra='forbid')
+
+class Locations(BaseModel):
```

Comment:
Why is Locations separated from ReconstructionSettings? Can we not just use reconstruction_dir: str in ReconstructionSettings?

Reply:
Ok this makes more sense now